### DISCUSSION TITLE
Derive selector of the entity wrapped by `EntityState`

### DISCUSSION THREAD

**Discussion by pBouillon** [state-adapt/state-adapt#62] [Q&A] [answered]
Hi,

I have defined an `adapter` for a `Player`, and its associated `EntityAdapter`:

```ts
export const playerAdapter = joinAdapters<Player>()({
  uuid: stringAdapter,
  score: scoreAdapter,
})({
  hasWon: (state: Player) => state.score.points === 50,
})();

export const playerEntityAdapter = createEntityAdapter<Player, PlayerId>()(
  playerAdapter,
);
```

From that definition, I created an `adapter` from a `Game` in which I would like to call the `hasWon` selector. This means that my `gameAdapter` is using `playerEntityAdapter` which relies on `playerAdapter`, which defines the selector I would like to use.

However, it seems that I can only access the value itself in `gameAdapter`'s selectors:

```ts
interface GameState {
  players: EntityState<Player, PlayerId>;
}

export const gameAdapter = joinAdapters<GameState>()({
  players: playerEntityAdapter,
})({
  winner: (state) =>
    state.playersAll.find(
      (player) => player.score.points === 50
      // 👆 logic already defined in playerAdapter.hasWon
    ) || null,
})();
```

How could I derive or compose the `playerAdapter.hasWon` selector?

**Comment by mfp22** [comment, marked as answer]
In the docs I used a config object to define filter selectors. https://state-adapt.github.io/adapters/core#createEntityAdapter

They have to return booleans. But this is how it applies to your code

```typescript 
export const playerEntityAdapter = createEntityAdapter<Player, PlayerId>()(
  playerAdapter,
  {
    filters: ['hasWon'],
  },
);
```

Now you can verify with TypeScript autocomplete, but I'm pretty sure playerEntityAdapter will now have a selector called `hasWon` and it will return an array of players that have won. So that should become gameAdapter.playersHasWon

Let me know if that works.

**Reply by mfp22** [reply]
Another approach that wouldn't be as fun would be to use the selector from playerAdapter directly again. The one created through the config object filters will actually memoize each player individually, but calling the selector directly wouldn't enable memoization. But these are concerns that wouldn't even be possible in another library, so I'm okay with where things are. Assuming this works for you.

**Reply by pBouillon** [reply]
It worked, thanks a lot, I think I simply didn't get how `filters` properly worked.

I updated the `EntityAdapter` as you said:

```ts
export const playerEntityAdapter = createEntityAdapter<Player, PlayerId>()(
  playerAdapter,
  {
    filters: ['hasWon'],
  },
);
```

And was able to derive the result in the parent `adapter`:

```ts
interface GameState {
  players: EntityState<Player, PlayerId>;
}

export const initialState: Readonly<GameState> = {
  players: createEntityState<Player, PlayerId, GameState['players']>(),
};

export const gameAdapter = joinAdapters<GameState>()({
  players: playerEntityAdapter,
})({
  winner: (state) => state.playersHasWon.at(0) || null,
})();
```

And I can consume it:

```ts
const state = adapt(initialState, {
  path: 'game',
  adapter: gameAdapter,
});

const winner$ = state.winner$;
```

Thanks a lot! 🎉

**Reply by mfp22** [reply]
Great, glad it worked. I really like this. Am I crazy, or is this really magical? I got so excited when I realized I could make this kind of thing happen. The grammar can be a bit awkward, but I like what you did defining a winner selector. The logic isn't repeated anywhere and it has a nice name at the end. `playersHasWon` isn't the worst name in the world I guess... Maybe it's pretty great actually. Any alternative I can think of would be worse.

Idk, but I'm excited to see more people diving into this.
